File: services/pkg/auth/permissions.go
package auth

import "strings"

// PermissionMatrix maps a cluster ID to the permission keys granted in that
// cluster. The special key "*" applies to every cluster — it carries the
// caller's global (admin.*) permissions and, for a full superuser, the "*"
// wildcard. Per-cluster resource/menu/ai.tool permissions live under the
// concrete cluster id and ONLY there (deny-by-default — no cluster fallback,
// 00-COMMON §2-3).
type PermissionMatrix map[string][]string

// ParsePermissions converts a raw JWT "permissions" claim into a matrix.
//
// Only the per-cluster map form is accepted. The legacy flat-array form (and a
// missing/!map claim) returns nil so the caller rejects the token with 401 and
// forces a one-time re-login after the format switch (00-COMMON §2-3). An empty
// map is a valid (deny-everything) matrix and is NOT rejected.
func ParsePermissions(raw any) PermissionMatrix {
	m, ok := raw.(map[string]any)
	if !ok {
		return nil
	}
	out := make(PermissionMatrix, len(m))
	for cid, v := range m {
		list, ok := v.([]any)
		if !ok {
			continue
		}
		perms := make([]string, 0, len(list))
		for _, p := range list {
			if s, ok := p.(string); ok {
				perms = append(perms, s)
			}
		}
		out[cid] = perms
	}
	return out
}

// HasForCluster reports whether perm is granted in clusterID. The all-cluster
// entry ("*") is always consulted; the concrete cluster's entry is consulted
// when clusterID is a real id. Matching reuses the wildcard rules: exact, full
// "*", and prefix globs like "resource.*".
func (m PermissionMatrix) HasForCluster(perm, clusterID string) bool {
	if matchAny(m["*"], perm) {
		return true
	}
	if clusterID != "" && clusterID != "*" {
		return matchAny(m[clusterID], perm)
	}
	return false
}

func matchAny(perms []string, perm string) bool {
	for _, pp := range perms {
		if permMatches(pp, perm) {
			return true
		}
	}
	return false
}

// permMatches compares a granted pattern with a required permission key by
// dot-separated segment: "*" alone grants everything; a "*" segment matches
// exactly one segment ("resource.*.read" → "resource.pod.read"); a trailing
// "*" matches the rest ("ai.tool.*" → "ai.tool.k8s_scale"). The seeded system
// roles rely on the middle form, so it must be honoured by every validator
// (Go here, ai-service security.py, frontend utils/permissions.ts).
func permMatches(pattern, perm string) bool {
	if pattern == "*" || pattern == perm {
		return true
	}
	ps := strings.Split(pattern, ".")
	qs := strings.Split(perm, ".")
	for i, seg := range ps {
		if seg == "*" && i == len(ps)-1 {
			return len(qs) >= len(ps)
		}
		if i >= len(qs) || (seg != "*" && seg != qs[i]) {
			return false
		}
	}
	return len(ps) == len(qs)
}

File: frontend/src/utils/permissions.ts
// Per-cluster permission matrix on the frontend (step 06/09).
//
// The authoritative source is the matrix GET /auth/me returns — the same one
// the backend puts in the JWT and enforces (services/pkg/auth). UI gating must
// use this rather than the global role's flat permission list, otherwise we'd
// show buttons the backend then 403s (a user's global role may include
// resource.* perms while their effective per-cluster access is deny-by-default
// until granted).

// clusterID → permission keys. The "*" entry holds GLOBAL (admin.*) perms and,
// for a superuser, the "*" wildcard.
export type PermissionMatrix = Record<string, string[]>

// permMatches mirrors pkg/auth permMatches (Go) and ai-service security.py:
// "*" grants everything; a "*" segment matches exactly one segment
// ("resource.*.read" → "resource.pod.read"); a trailing "*" matches the rest
// ("ai.tool.*" → "ai.tool.k8s_scale").
export function permMatches(pattern: string, perm: string): boolean {
  if (pattern === '*' || pattern === perm) return true
  const ps = pattern.split('.')
  const qs = perm.split('.')
  for (let i = 0; i < ps.length; i++) {
    if (ps[i] === '*' && i === ps.length - 1) return qs.length >= ps.length
    if (i >= qs.length || (ps[i] !== '*' && ps[i] !== qs[i])) return false
  }
  return ps.length === qs.length
}

export function matchAny(perms: string[] | undefined, perm: string): boolean {
  if (!perms) return false
  return perms.some((p) => permMatches(p, perm))
}

// hasPermission honors the all-cluster "*" entry always; the concrete cluster's
// entry is consulted when clusterID is a real id (deny-by-default — no fallback).
export function hasPermission(matrix: PermissionMatrix, perm: string, clusterID = ''): boolean {
  if (matchAny(matrix['*'], perm)) return true
  return clusterID ? matchAny(matrix[clusterID], perm) : false
}

// parsePermissions accepts only the per-cluster map form. A flat array or a
// missing/!object value yields an empty matrix (deny-all), matching the
// backend's reject-and-relogin behavior on the UI side.
export function parsePermissions(raw: unknown): PermissionMatrix {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return {}
  const out: PermissionMatrix = {}
  for (const [cid, v] of Object.entries(raw as Record<string, unknown>)) {
    if (Array.isArray(v)) {
      out[cid] = v.filter((p): p is string => typeof p === 'string')
    }
  }
  return out
}
